happi/item.py

```diff
@@ -0,0 +1,392 @@
+    """
+    The smallest description of an object that can be entered in ``happi``
+
+    The class does not need to be intialized with any specific piece of
+    information except a name, but all of the attributes listed by
+    :attr:`HappiItem.info_names` can be used to assign values to
+    :class:`.EntryInfo` upon initialization.  Pieces of information that are
+    deemed mandatory by the class must be filled in before the device is loaded
+    into the database. See :attr:`.mandatory_info` to see which
+    attributes are neccesary.
+
+    Additional metadata can be given to the device in the form of keywords on
+    initialization, this information is kept in the :attr:`.extraneous`
+    attribute, and will be saved in to the database as long as it does not
+    clash with an existing piece of metadata that the client uses to organize
+    devices.
+
+    Attributes
+    ----------
+    entry_info : list
+        A list of all the :class:`.EntryInfo` associated with the device
+
+    extraneous : dict
+        Storage for information supplied during initialization that does not
+        correspond to a specific EntryInfo
+
+    Raises
+    ------
+    EntryError:
+        If a piece of information supplied at startup is of the incorrect type
+
+    ContainerError:
+        If one of the pieces of :class:`.EntryInfo` has a default value of the
+        incorrect type
+
+    Example
+    -------
+    .. code ::
+
+        d = HappiItem(name = 'my_device',         # Alias name for device
+                      note  = 'Example',          # Piece of arbitrary metadata
+                     )
+    """
+    name = EntryInfo('Shorthand Python-valid name for the Python instance',
+                     optional=False,
+                     enforce=re.compile(r'[a-z][a-z\_0-9]{2,78}$'))
+    device_class = EntryInfo("Python class that represents the instance",
+                             enforce=str)
+    args = EntryInfo("Arguments to pass to device_class",
+                     enforce=list, default=[])
+    kwargs = EntryInfo("Keyword arguments to pass to device_class",
+                       enforce=dict, default={})
+    active = EntryInfo('Whether the instance is actively deployed',
+                       enforce=bool, default=True)
+    documentation = EntryInfo("Relevant documentation for the instance",
+                              enforce=str)
+
+    def __init__(self, **kwargs):
+        # Load given information into device class
+        for info in self.entry_info:
+            if info.key in kwargs.keys():
+                setattr(self, info.key, kwargs.pop(info.key))
+            else:
+                setattr(self, info.key, info.default)
+        # Handle additional information
+        self.extraneous = kwargs
+        if kwargs:
+            logger.debug('Additional information for %s was defined %s',
+                         self.name, ', '.join(self.extraneous))
+
+    @property
+    def info_names(self):
```

Comment:
We should move these into the `_HappiItem` `__init_subclass__`, as `info_names`, `mandatory_info` does not change.

Reply:
Done